src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+#include <ctype.h>
+#include <dirent.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <sol-log.h>
+#include <sol-mainloop.h>
+#include <sol-str-slice.h>
+#include <sol-util.h>
+#include <sol-util-linux.h>
+#include <sol-vector.h>
+
+#include "sol-iio.h"
+
+struct sol_iio_device {
+    char *trigger_name;
+    sol_iio_reader_cb reader_cb;
+    void *reader_cb_data;
+    struct sol_fd *fd_handler;
+    struct sol_ptr_vector channels;
+    int device_id;
+    int trigger_id;
+    int fd;
+    bool buffer_enabled : 1;
+    bool manual_triggering : 1;
+};
+
+struct sol_iio_channel {
+    struct sol_iio_device *device;
+    char *name;
+    double scale;
+    int index;
+    int offset;
+
+    int storagebits;
+    int bits;
+    int shift;
+    int offset_in_buffer;
+
+    bool little_endian : 1;
+    bool is_signed : 1;
+};
+
+struct sol_iio_buffer {
+    size_t size;
+    uint8_t *data;
+};
```

Comment:
I'd change this to `struct sol_buffer`, it contains an extra `reserved` field that we don't need, but saves us from defining yet-another structure.